scripts/start.js

```diff
@@ -66,62 +69,90 @@ function clearConsole() {
 }
 
 var compiler = webpack(config, handleCompile);
-compiler.plugin('invalid', function () {
-  clearConsole();
-  console.log('Compiling...');
-});
-compiler.plugin('done', function (stats) {
-  clearConsole();
-  var hasErrors = stats.hasErrors();
-  var hasWarnings = stats.hasWarnings();
-  if (!hasErrors && !hasWarnings) {
-    console.log(chalk.green('Compiled successfully!'));
-    console.log();
-    console.log('The app is running at http://localhost:3000/');
-    console.log();
-    return;
-  }
+detect(PORT, (error, _port) => {
 
-  var json = stats.toJson();
-  var formattedErrors = json.errors.map(message =>
-    'Error in ' + formatMessage(message)
-  );
-  var formattedWarnings = json.warnings.map(message =>
-    'Warning in ' + formatMessage(message)
-  );
+  if (PORT !== _port) {
+    var rl = readline.createInterface({
+      input: process.stdin,
+      output: process.stdout
+    });
 
-  if (hasErrors) {
-    console.log(chalk.red('Failed to compile.'));
-    console.log();
-    if (formattedErrors.some(isLikelyASyntaxError)) {
-      // If there are any syntax errors, show just them.
-      // This prevents a confusing ESLint parsing error
-      // preceding a much more useful Babel syntax error.
-      formattedErrors = formattedErrors.filter(isLikelyASyntaxError);
-    }
-    formattedErrors.forEach(message => {
-      console.log(message);
-      console.log();
+    rl.question('Something is already running at http://localhost:' + PORT + '. Would you like to run the app at another port instead? [Y/n] ', (answer) => {
+      if(answer === 'Y') {
+        PORT = _port;
+        // Replace the port in webpack config
+        config.entry = config.entry.map(c => c.replace(/(\d+)/g, PORT));
+        compiler = webpack(config, handleCompile);
+        setupCompiler();
```

Comment:
It’s bad that we repeat this code in this and the other branch below. This makes it easy to change something in one place and forget to change it in the other one.

Instead, let’s split the logic into something like

```js
function choosePort() {
  return new Promise(resolve => {
    detect(...)
  })
}
```

and then do something like

```js
choosePort().then(port => {
  setupCompiler(port);
  ...
});
```